parser_library/src/parsing/parser_error_listener.cpp

```diff
@@ -114,135 +115,155 @@ void iterate_error_stream(antlr4::TokenStream* input_stream,
+void parser_error_listener_base::handle_error(token_stream* input_stream,
+    int start_index,
+    int end_index,
+    size_t line,
+    size_t char_pos_in_line,
+    const antlr4::Token* start_token,
+    const antlr4::misc::IntervalSet& expected_tokens)
+{
+    bool sign_followed = true;
+    bool sign_preceding = true;
+    bool only_par = true;
+    bool right_prec = false;
+    bool left_prec = false;
+    bool unexpected_sign = false;
+    bool odd_apostrophes = false;
+    bool ampersand_followed = true;
 
-        // while it's a space, skip spaces
-        while (first_symbol_type == SPACE)
-        {
-            start_index++;
-            first_symbol_type = input_stream->get(start_index)->getType();
-        }
+    iterate_error_stream(input_stream,
+        start_index,
+        end_index,
+        right_prec,
+        only_par,
+        left_prec,
+        sign_followed,
+        sign_preceding,
+        unexpected_sign,
+        odd_apostrophes,
+        ampersand_followed);
 
-        auto end_index = get_end_index(input_stream, start_index);
+    // ampersand not followed with a name of a variable symbol
+    if (!ampersand_followed)
+        add_parser_diagnostic(diagnostic_op::error_S0008, range(position(line, char_pos_in_line)));
+    // apostrophe expected
+    else if (odd_apostrophes && expected_tokens.contains((size_t)APOSTROPHE))
+        add_parser_diagnostic(diagnostic_op::error_S0005, range(position(line, char_pos_in_line)));
+    // right parenthesis has no left match
+    else if (right_prec)
+        add_parser_diagnostic(diagnostic_op::error_S0012, range(position(line, char_pos_in_line)));
+    // left parenthesis has no right match
+    else if (left_prec)
+        add_parser_diagnostic(diagnostic_op::error_S0011, range(position(line, char_pos_in_line)));
+    // nothing else but left and right parenthesis is present
+    else if (only_par)
+        add_parser_diagnostic(diagnostic_op::error_S0010, range(position(line, char_pos_in_line)));
+    // sign followed by a wrong token
+    else if (!sign_followed)
+        add_parser_diagnostic(diagnostic_op::error_S0009, range(position(line, char_pos_in_line)));
+    // expression starting with a sign
+    else if (!sign_preceding)
+        add_parser_diagnostic(diagnostic_op::error_S0007, range(position(line, char_pos_in_line)));
+    // unexpected sign in an expression - GT, LT etc
+    else if (unexpected_sign)
+        add_parser_diagnostic(diagnostic_op::error_S0006, range(position(line, char_pos_in_line)));
+    // unfinished statement - solo label on line
+    else if (start_token->getCharPositionInLine() == 0)
+        add_parser_diagnostic(diagnostic_op::error_S0004, range(position(line, char_pos_in_line)));
+    // other undeclared errors
+    else if (start_token->getType() == antlr4::Token::EOF)
+        add_parser_diagnostic(diagnostic_op::error_S0003, range(position(line, char_pos_in_line)));
```

Comment:
Maybe
```suggestion
    // EOF errors
    else if (start_token->getType() == antlr4::Token::EOF)
        add_parser_diagnostic(diagnostic_op::error_S0003, range(position(line, char_pos_in_line)));
```

?

Also, it seems like `S0003` is not used anymore? Or at least I didn't find any reference to it in the tests

Reply:
https://github.com/BroadcomMFD-DevOps/che-che4z-lsp-for-hlasm/blob/41d2f80850bc52caa12d4ceb0177a1bac3f05756/parser_library/test/parsing/string_test.cpp#L308